src/main/scala/org/apache/spark/sql/delta/catalog/DeltaCatalog.scala

```diff
@@ -237,51 +272,54 @@ class DeltaCatalog(val spark: SparkSession) extends DelegatingCatalogExtension
       ident: Identifier,
       override val schema: StructType,
       val partitions: Array[Transform],
-      override val properties: util.Map[String, String]) extends StagedTable with SupportsWrite {
+      override val properties: util.Map[String, String],
+      operation: TableCreationModes.CreationMode) extends StagedTable with SupportsWrite {
     override def name(): String = ident.name()
 
     override def abortStagedChanges(): Unit = {}
 
-    override def commitStagedChanges(): Unit = {}
+    private var asSelectQuery: Option[DataFrame] = None
+    private var writeOptions: Map[String, String] = properties.asScala.toMap
+
+    override def commitStagedChanges(): Unit = {
+      createDeltaTable(
+        ident,
+        schema,
+        partitions,
+        writeOptions.asJava,
+        asSelectQuery.map(_.logicalPlan),
+        operation)
+    }
 
     override def capabilities(): util.Set[TableCapability] = Set(
-      ACCEPT_ANY_SCHEMA, BATCH_READ,
-      V1_BATCH_WRITE, OVERWRITE_BY_FILTER, TRUNCATE
+      ACCEPT_ANY_SCHEMA, BATCH_READ, V1_BATCH_WRITE, OVERWRITE_BY_FILTER, TRUNCATE
     ).asJava
 
     override def newWriteBuilder(options: CaseInsensitiveStringMap): V1WriteBuilder = {
-      // TODO: is this right? What's the appropriate distinction to keep for properties and options
       val combinedProps = options.asCaseSensitiveMap().asScala ++ properties.asScala
-      new DeltaV1WriteBuilder(ident, schema, partitions, combinedProps.asJava)
+      writeOptions = combinedProps.toMap
+      new DeltaV1WriteBuilder
     }
-  }
 
-  /*
-   * We have to do extend both classes. Only extending V1WriteBuilder gives
-   *
-   * Unable to implement a super accessor required by trait V1WriteBuilder unless
-   * org.apache.spark.sql.sources.v2.writer.WriteBuilder is directly extended by class
-   * DeltaCatalog$DeltaV1WriteBuilder.
-   */
-  private class DeltaV1WriteBuilder(
-      ident: Identifier,
-      schema: StructType,
-      partitions: Array[Transform],
-      properties: util.Map[String, String]) extends WriteBuilder with V1WriteBuilder {
-    override def buildForV1Write(): InsertableRelation = {
-      new InsertableRelation {
-        override def insert(data: DataFrame, overwrite: Boolean): Unit = {
-          createDeltaTable(ident, schema, partitions, properties, Some(data.logicalPlan))
+    private class DeltaV1WriteBuilder extends WriteBuilder with V1WriteBuilder {
+      override def buildForV1Write(): InsertableRelation = {
+        new InsertableRelation {
+          override def insert(data: DataFrame, overwrite: Boolean): Unit = {
+            asSelectQuery = Some(data)
```

Comment:
nit: This action-at-a-distance is confusing. I don't think there's a better way to do it but I'd add a comment explaining.